287 Rip out guava

gnl42 opened the pull request:
Almost made it. Still a couple of uses left that I don't know how to replace:

1. TypeToken@GerritClient212
    Can it be replaced with com.google.gson.reflect.TypeToken?
2. MediaType/Multimap@TaskEditorExtensions
   Can it be replaced by javax.ws.rs.core.MediaType?
3. HashBasedTable/Table@ConnectorMigrator,BrandManager
    Doesn't seem to be a drop in alternative. 

@BeckerFrank @ruspl-afed How do I get Caffeine added to a P2 repository (given the hassle we had with maven artifacts in Jenkins)?

merks reviewed (approved):
Getting rid of this dependency is a good idea.

One word of caution though is that Set.of and List.of create collections that throw NPEs if one does a contains(null)...

ruspl-afed reviewed (changes requested):
@gnl42 Please remove all the unrelated changes like subversion addition from this PR

Moreover I would suggest to keep only these changes that doesn't involve new libraries instead of guava so we can merge the major part of your work.

merks commented:
If you need something in orbit you can ask with details here:

https://github.com/eclipse-orbit/orbit-simrel/issues

gnl42 commented on org.eclipse.mylyn.releng/launcher/org.eclipse.mylyn-aggregator threads.launch:
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Hmm, that wasn't supposed to be committed. 
It's so that I could speed up the Mylyn build by multi-threading it

gnl42 commented on org.eclipse.mylyn-target/org.eclipse.mylyn.target.target:
```diff
@@ -84,5 +87,20 @@
 			<unit id="com.google.gson" version="0.0.0"/>
 			<unit id="org.mockito.mockito-core" version="0.0.0"/>
 		</location>
+		<location includeAllPlatforms="false" includeConfigurePhase="false" includeMode="planner" includeSource="true" type="InstallableUnit">
```
It's not :( Merging with main re-introduced it

gnl42 commented:
> If you need something in orbit you can ask with details here:
> 
> https://github.com/eclipse-orbit/orbit-simrel/issues

Hi @merks, I had a look at the site when the orbit email went but I can't find any instructions/description/example on how to request/add a dependency.
I can do the work but I have no idea where to start

netomi commented:
Not sure how the reviewers have been assigned, but be aware that there are also members in the organization that are there to perform other duties than development, so be mindful with assigning members as reviewers.

gnl42 commented:
@netomi 
> Not sure how the reviewers have been assigned, but be aware that there are also members in the organization that are there to perform other duties than development, so be mindful with assigning members as reviewers.

Sorry, I assumed that the reviewers available for the PR were associated with the project.
I'll clean up the list

merks commented:
Unfortunately documentation is still TBD.  In the end, I just need the dependency information `groupId:artifactId:artifactVersion` spelled out in an issue.

I've spent some time already looking at  Mylyn's dependencies and there is bunch of really old stuff that is not in the latest Orbit.  😱

gnl42 commented:
@merks 
> Getting rid of this dependency is a good idea.
> 
> One word of caution though is that Set.of and List.of create collections that throw NPEs if one does a contains(null)...

guava seems to be the same from what I can tell

merks commented:
> @merks
> 
> > Getting rid of this dependency is a good idea.
> > One word of caution though is that Set.of and List.of create collections that throw NPEs if one does a contains(null)...
> 
> guava seems to be the same from what I can tell

That's perfect then!  Thanks for double checking.

wimjongman commented on mylyn.context/org.eclipse.mylyn.context.tasks.tests/src/org/eclipse/mylyn/context/tasks/tests/TaskEditorRestoreTest.java:
```diff
@@ -156,19 +152,27 @@ public String apply(TextEditor editor) {
+//				})
+//				.filter(clazz)
+//				.transform(propertyFunction)
+//				.toSet();
```
The comments can go. Git knows.

gnl42 replied:
Gone

wimjongman commented on mylyn.context/org.eclipse.mylyn.context.tasks.ui/src/org/eclipse/mylyn/internal/context/tasks/ui/TaskActivityMonitor.java:
```diff
@@ -111,13 +108,16 @@ private void activateEditors(List<IEditorReference> refs) {
+//				public String apply(IEditorReference ref) {
+//					return ref.getTitle();
+//				}
+//			}));
```
Comments can go.

gnl42 replied:
Gone

wimjongman commented on mylyn.reviews/org.eclipse.mylyn.gerrit.core/src/org/eclipse/mylyn/internal/gerrit/core/GerritConnector.java:
```diff
@@ -358,19 +359,22 @@ private void addExtendedTooltip(ITask task, TaskAttribute projectAttribute) {
-				.join(projectTooltip, branchTooltip, reviewTooltip, verifiedTooltip);
+//		String tooltip = Joiner.on("\n") //$NON-NLS-1$
+//				.skipNulls()
+//				.join(projectTooltip, branchTooltip, reviewTooltip, verifiedTooltip);
```
Comments can go

gnl42 replied:
Gone

wimjongman commented on mylyn.reviews/org.eclipse.mylyn.gerrit.core/src/org/eclipse/mylyn/internal/gerrit/core/GerritTaskDataHandler.java:
```diff
@@ -297,10 +297,19 @@ public void updateTaskData(TaskRepository repository, TaskData data, GerritChang
+//		}
+//
+//
+//		for (Entry<Integer, Collection<BuildResult>> buildResult : resultsByPatchNumber.asMap().entrySet()) {
```
More comments

gnl42 replied:
Gone

wimjongman commented on mylyn.reviews/org.eclipse.mylyn.gerrit.core/src/org/eclipse/mylyn/internal/gerrit/core/client/GerritClient.java:
```diff
@@ -780,15 +778,18 @@ private void cacheBranches(Project.NameKey project, IProgressMonitor monitor) th
+//						return input.getRef();
+//					}
+//				})
+//				.toSet();
```
Commented code

gnl42 replied:
Gone

wimjongman commented on mylyn.reviews/org.eclipse.mylyn.gerrit.core/src/org/eclipse/mylyn/internal/gerrit/core/remote/GerritReviewRemoteFactory.java:
```diff
@@ -250,20 +247,22 @@ public void updateParentCommits(IRepository parent, IReview review, ChangeDetail
+//			public boolean apply(IReviewItemSet set) {
+//				return set.getId().equals(patchSetNumber);
+//			}
+//		}).orNull();
```
Commented code

wimjongman commented on mylyn.reviews/org.eclipse.mylyn.gerrit.core/src/org/eclipse/mylyn/internal/gerrit/core/remote/GerritReviewRemoteFactory.java:
```diff
@@ -250,20 +247,22 @@ public void updateParentCommits(IRepository parent, IReview review, ChangeDetail
+//				return commit.getId().equals(candidateCommit.getId());
+//			}
+//		});
+//		return optional.isPresent();
```
Commented code

wimjongman commented on mylyn.reviews/org.eclipse.mylyn.gerrit.tests/src/org/eclipse/mylyn/internal/gerrit/core/remote/GerritReviewRemoteFactoryJUnit3Test.java:
```diff
@@ -694,11 +692,17 @@ public void testGlobalCommentByGerrit() throws Exception {
+//			public boolean apply(IUser user) {
+//				return user.getDisplayName().equals(displayName);
+//			}
+//		});
```
Commented code

wimjongman commented on mylyn.reviews/org.eclipse.mylyn.gerrit.ui.tests/src/org/eclipse/mylyn/internal/gerrit/ui/editor/GerritTaskEditorPageTest.java:
```diff
@@ -100,20 +95,25 @@ protected void setUp() throws Exception {
+//							public Class<?> apply(TaskEditorPartDescriptor o) {
+//								return o.createPart().getClass();
+//							}
+//						}));
```
Commented code

wimjongman commented on mylyn.reviews/org.eclipse.mylyn.gerrit.ui.tests/src/org/eclipse/mylyn/internal/gerrit/ui/editor/GerritTaskEditorPageTest.java:
```diff
@@ -100,20 +95,25 @@ protected void setUp() throws Exception {
+//			public String apply(TaskEditorPartDescriptor o) {
+//				return o.getId();
+//			}
+//		});
```
Commented code

wimjongman commented on mylyn.reviews/org.eclipse.mylyn.gerrit.ui.tests/src/org/eclipse/mylyn/internal/gerrit/ui/editor/GerritTaskEditorPageTest.java:
```diff
@@ -138,20 +138,22 @@ public void testRemoveUnneededSections() {
+//			public boolean apply(TaskEditorPartDescriptor descriptor) {
+//				return descriptor.getPath().equals(path);
+//			}
+//		});
```
Commented code

gnl42 replied:
Gone

wimjongman commented on mylyn.reviews/org.eclipse.mylyn.gerrit.ui.tests/src/org/eclipse/mylyn/internal/gerrit/ui/editor/GerritTaskEditorPageTest.java:
```diff
@@ -138,20 +138,22 @@ public void testRemoveUnneededSections() {
+//			public boolean apply(TaskEditorPartDescriptor descriptor) {
+//				return descriptor.getId().equals(id);
+//			}
+//		});
```
Commented code

gnl42 replied:
Gone

wimjongman commented on mylyn.reviews/org.eclipse.mylyn.gerrit.ui/src/org/eclipse/mylyn/internal/gerrit/ui/GerritCompareUi.java:
```diff
@@ -81,27 +81,47 @@ private static CompareEditorInput getComparisonEditor(CompareEditorInput editorI
+//				}
+//				return null;
+//			}
+//		}).firstMatch(predicate).or(editorInput);
```
Commented code

gnl42 replied:
Gone

wimjongman commented on mylyn.reviews/org.eclipse.mylyn.gerrit.ui/src/org/eclipse/mylyn/internal/gerrit/ui/GerritUrlHandler.java:
```diff
@@ -127,7 +125,8 @@ public String getFilePath(TaskRepository repository, String url, String taskId,
 			String[] fragments = StringUtils.split(extractUrlQualifiers(repository, url, taskId), "/"); //$NON-NLS-1$
 			if (fragments.length > 1) {
-				return Joiner.on("/").join(Arrays.copyOfRange(fragments, 1, fragments.length)); //$NON-NLS-1$
+//				return Joiner.on("/").join(Arrays.copyOfRange(fragments, 1, fragments.length)); //$NON-NLS-1$
```
Commented code

gnl42 replied:
Gone

wimjongman commented on mylyn.reviews/org.eclipse.mylyn.gerrit.ui/src/org/eclipse/mylyn/internal/gerrit/ui/factories/CompareWithUiFactory.java:
```diff
@@ -103,11 +101,13 @@ public void updated(boolean modified) {
+//					public String apply(IFileItem f) {
+//						return f.getName();
+//					}
+//				}));
```
Commented code

gnl42 replied:
Gone

wimjongman reviewed (changes requested):
George, do you want to clean up the commented code first? I made it halfway through.

gnl42 commented:
@wimjongman 
> George, do you want to clean up the commented code first? I made it halfway through.

I my thought is/was to leave the commented out code in for a while.
Both to server as a reference for the reviewers so they could validate the new code. And as a reference if something goes wrong when the code is released. If the comments are removed it would make it harder to track down a problem since it would appear that nothing has changed.

I was thinking waiting til' something like 4.2.0 to clean up the code

gnl42 commented:
Ok, this is annoying. While addressing the PR comments I happened to be looking at the console just as this went by:
```
Running org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUiTest
Tests run: 5, Failures: 0, Errors: 5, Skipped: 0, Time elapsed: 0.234 s <<< FAILURE! - in org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUiTest
org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUiTest.testOpenNewFileCompareEditor  Time elapsed: 0.203 s  <<< ERROR!
java.lang.NoSuchMethodError: 'com.google.common.base.Predicate org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUi.getFileComparePredicate(org.eclipse.mylyn.reviews.core.model.IFileItem)'
	at org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUiTest.testOpenNewFileCompareEditor(GerritCompareUiTest.java:106)

org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUiTest.testOpenNewReviewItemSetCompareEditor  Time elapsed: 0.025 s  <<< ERROR!
java.lang.NoSuchMethodError: 'com.google.common.base.Predicate org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUi.getReviewItemSetComparePredicate(org.eclipse.mylyn.reviews.core.model.IReviewItemSet, java.lang.String)'
	at org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUiTest.testOpenNewReviewItemSetCompareEditor(GerritCompareUiTest.java:117)

org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUiTest.testOpenExistingFileCompareEditor  Time elapsed: 0.002 s  <<< ERROR!
java.lang.NoSuchMethodError: 'com.google.common.base.Predicate org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUi.getFileComparePredicate(org.eclipse.mylyn.reviews.core.model.IFileItem)'
	at org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUiTest.testOpenExistingFileCompareEditor(GerritCompareUiTest.java:135)

org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUiTest.testNoOpenCompareEditors  Time elapsed: 0.001 s  <<< ERROR!
java.lang.NoSuchMethodError: 'com.google.common.base.Predicate org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUi.getFileComparePredicate(org.eclipse.mylyn.reviews.core.model.IFileItem)'
	at org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUiTest.testNoOpenCompareEditors(GerritCompareUiTest.java:91)

org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUiTest.testOpenExistingReviewItemSetCompareEditor  Time elapsed: 0.002 s  <<< ERROR!
java.lang.NoSuchMethodError: 'com.google.common.base.Predicate org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUi.getReviewItemSetComparePredicate(org.eclipse.mylyn.reviews.core.model.IReviewItemSet, java.lang.String)'
	at org.eclipse.mylyn.internal.gerrit.ui.GerritCompareUiTest.testOpenExistingReviewItemSetCompareEditor(GerritCompareUiTest.java:147)

I can't tell where the hidden reference to Predicate is coming from since there is no stack trace.
When I run the same test under Eclipse it passes with no problems.

How can I get some more information about what is happening here?

Also, the maven build passes with no errors. Why is this error being swallowed by the build process?

@merks @wimjongman @ruspl-afed @merks 
```

ruspl-afed commented:
@gnl42  

> Ok, this is annoying.

Who said guava would be an easy frag?
Generally `java.lang.NoSuchMethodError` indicates that method was present during compilation but cannot be found during runtime. You can try to explicitly add dependency to guava to the failing test bundle until we remove guava completely.

> Also, the maven build passes with no errors. Why is this error being swallowed by the build process?

Do you mean that locally you have no such issues when doing
```
mvn clean verify
```
?

ruspl-afed commented:
> George, do you want to clean up the commented code first? 

I agree with @wimjongman, Mylyn codebase is already mysterious enough to add commented out code.
We either replace Guava calls with equivalents or keep them as they are, and we'll let Git keep the history for us.

gnl42 commented:
I'll removed the guava code, if I can figure out why a unit test works for me and fails with jenkins :(

akurtakov commented:
When I removed guava from a number of projects I figured it's easier to do it multiple steps - first the direct replacements from jdk then the more advanced ones and let them land in master so it was easier to figure what happens. I hope that helps.

merks commented:
When I look at the PR in my workspace, it's clear you missed one bundle and worse yet, guava is still in the target platform so nothing will fail to build simply because it's missing.

![image](https://github.com/eclipse-mylyn/org.eclipse.mylyn/assets/208716/02243001-481f-4083-a3c7-331edbed9ed7)

When I remove it from the target platform, I see all these errors:

![image](https://github.com/eclipse-mylyn/org.eclipse.mylyn/assets/208716/f5ffc3ff-4daa-44b8-9ba3-467a77ff6963)

So there are also package imports that have been missed in this guava-ectomy.

-----

I hope Mylyn will build against 2023-12 Orbit not 2023-09 for the milestone:

![image](https://github.com/eclipse-mylyn/org.eclipse.mylyn/assets/208716/ca894426-3682-4a07-be24-8a9ac9e333ab)

gnl42 commented:
@merks There are still about four classes that uses guava.

I'm don't know why you are seeing all those errors. My instance is error clean
![image](https://github.com/eclipse-mylyn/org.eclipse.mylyn/assets/66577321/c651650c-64e4-47b7-b9f5-db4edc07c9cd)

merks commented:
As I mentioned, in your PR guava is still present in the target platform.   My errors come when I remove it from the target platform like this:

![image](https://github.com/eclipse-mylyn/org.eclipse.mylyn/assets/208716/5507a448-3d18-44a2-9805-cbd1a5d4df82)

Then I used Perform Setup Tasks from the toolbar and perform that one task underlined above to re-resolve the target platform.  Then all the things using guava have errors and these are what needs to be cleaned up further to completely remove guava.

I'm not sure it makes sense to try to figure out why the build does not work in the current intermediate state of partial removal...

gnl42 commented:
As I said, there are still a few guava uses left on the code, primarily because of:
- Are the alternatives drop in replacements for: TypeToken
- Are the alternatives compatible with: MediaType
- No drop-in alternatives. Would require code/logic change: HashBasedTable
- Unit tests fail due to results in different order. Don't know if order is significant

As it is, at the moment the best that can be done is to minimize any issues with change guava version.
We can however wipe most of guava away.

I'm up for any suggestions where to go from here.
- Forget the whole thing
- Staged PRs
- Continue with big bang approach

merks commented:
Personally I'm fine with the big bang approach.  I think removing it, if possible, will save long term hassles because this library loves to create new major versions on a regular basis and multiple installed versions of it are the single biggest cause of OSGi wiring problems.

wimjongman commented:
>     * Continue with big bang approach

+1 for this option as well. Thanks for working on this, George.

gnl42 commented:
Finally have clean build with no unit errors. I'll clean up the commented out code now.

Realized that guava ImmutableMap preserves insert order (duh!??). Two of the unit tests required a specific order of results.
The question is, will not preserving the insert order cause any issues in real life?

Some other questions: How on earth do the unit tests work?
- We have unit tests that pass, yet there is a stack trace with explicit assertXXX failures that somehow are completely ignored.
- There are unit tests that can't possibly work: method is passed null, first thing method does it reject a null argument.
- Why is does org.eclipse.mylyn-aggregator.launch set 'maven.test.failure.ignore=true'? Why are we ignoring test errors?

gnl42 commented:
Unit tests all pass and commented out gitcode removed.

@wimjongman, @ruspl-afed Can you approved the PR now so it can be merged

wimjongman commented:
Thanks, George. I have approved the changes.

ruspl-afed reviewed (approved):
Outstanding job @gnl42 !

I have some minor comments, but let's merge this PR first.
